File: go/fibonacci.go
package main

func Fibonacci(n int) int {
	if n < 2 {
		return n
	}
	return Fibonacci(n-1) + Fibonacci(n-2)
}

type Memo map[int]int

func MemoizedFibonacci(n int, cache Memo) int {
	if n < 2 {
		return n
	}
	if _, ok := cache[n]; !ok {
		cache[n] = MemoizedFibonacci(n-1, cache) + MemoizedFibonacci(n-2, cache)
	}
	return cache[n]
}

File: js/fibonacci.ts
export function fibonacci(n: number): number {
    if (n < 2) {
        return n;
    }

    return fibonacci(n - 1) + fibonacci(n - 2);
}

export function memoizedFibonacci(n: number, cache: { [key: number]: number } = {}): number {
    if (n < 2) {
        return n;
    }
  
    if (!(n in cache)) {
      cache[n] = memoizedFibonacci(n - 1, cache) + memoizedFibonacci(n - 2, cache);
    }

    return cache[n];
}
